homeassistant/components/tplink/entity.py

```diff
@@ -62,24 +108,278 @@ async def _async_wrap(self: _T, *args: _P.args, **kwargs: _P.kwargs) -> None:
+        if key == PRIMARY_STATE_ID:
+            return legacy_device_id(self._device)
+
+        # Historically the legacy device emeter attributes which are now replaced with
+        # features used slightly different keys. This ensures that those entities
+        # are not orphaned. Returns the mapped key or the provided key if not mapped.
+        key = LEGACY_KEY_MAPPING.get(key, key)
+        return f"{legacy_device_id(self._device)}_{key}"
+
+    @classmethod
+    def _category_for_feature(cls, feature: Feature | None) -> EntityCategory | None:
+        """Return entity category for a feature."""
+        # Main controls have no category
+        if feature is None or feature.category is Feature.Category.Primary:
+            return None
+
+        if (
+            entity_category := FEATURE_CATEGORY_TO_ENTITY_CATEGORY.get(feature.category)
+        ) is None:
+            _LOGGER.error(
+                "Unhandled category %s, fallback to DIAGNOSTIC", feature.category
+            )
+            entity_category = EntityCategory.DIAGNOSTIC
+
+        return entity_category
+
+    @classmethod
+    def _description_for_feature[_D: EntityDescription](
+        cls,
+        desc_cls: type[_D],
+        feature: Feature,
+        descriptions: Mapping[str, _D],
+        *,
+        child_alias: str | None = None,
+    ) -> _D:
+        """Return description object for the given feature.
+
+        This is responsible for setting the common parameters & deciding based on feature id
+        which additional parameters are passed.
+        """
+
+        enabled_default = feature.category is not Feature.Category.Debug
+        category = cls._category_for_feature(feature)
+        if descriptions and (desc := descriptions.get(feature.id)):
+            translation_key: str | None = feature.id
+            # HA logic is to name entities based on the following logic:
+            # _attr_name > translation.name > description.name
+            # > device_class (if base platform supports).
+            entity_name: str | None | UndefinedType = UNDEFINED
+
+            # The state feature gets the device name or the child device
+            # name if it's a child device
+            if feature.id == PRIMARY_STATE_ID:
+                translation_key = None
+                entity_name = child_alias  # if None will use device name
+
+            desc = replace(
+                desc,
+                translation_key=translation_key,
+                name=entity_name,  # if undefined will use translation key
+                entity_category=category,
+                entity_registry_enabled_default=enabled_default,
+            )
+        else:
+            desc = desc_cls(
+                key=feature.id,
+                name=feature.name,
+                icon=feature.icon,
+                entity_category=category,
+                entity_registry_enabled_default=enabled_default,
+            )
```

Comment:
```suggestion
```
We should avoid creating arbitrary descriptions for features not specifically supported or considered in HA so we don't have to clean this up later/try to handle issues about half working functionality.  Its better its not there at all vs not working properly.

related https://github.com/home-assistant/core/pull/117760/files#r1649286726

Reply:
Fixed in https://github.com/home-assistant/core/pull/120141